src/serialization/ParseNodeFactoryRegistry.cs

```diff
@@ -54,6 +58,43 @@ public IParseNode GetRootParseNode(string contentType, Stream content)
             if(ContentTypeAssociatedFactories.TryGetValue(cleanedContentType, out var factory))
                 return factory.GetRootParseNode(cleanedContentType, content);
 
+            throw new InvalidOperationException($"Content type {cleanedContentType} does not have a factory registered to be parsed");
+        }
+        /// <summary>
+        /// Get the <see cref="IParseNode"/> instance that is the root of the content
+        /// </summary>
+        /// <param name="contentType">The content type of the stream</param>
+        /// <param name="content">The <see cref="Stream"/> to parse</param>
+        /// <param name="cancellationToken">The cancellation token for the task</param>
+        /// <returns></returns>
+        public async Task<IParseNode> GetRootParseNodeAsync(string contentType, Stream content, 
+            CancellationToken cancellationToken = default)
+        {
+            if(string.IsNullOrEmpty(contentType))
+                throw new ArgumentNullException(nameof(contentType));
+            _ = content ?? throw new ArgumentNullException(nameof(content));
+
+            var vendorSpecificContentType = contentType.Split(";".ToCharArray(), StringSplitOptions.RemoveEmptyEntries).First();
+            if(ContentTypeAssociatedFactories.TryGetValue(vendorSpecificContentType, out var vendorFactory))
+            {
+                if(vendorFactory is not IAsyncParseNodeFactory vendorFactoryAsync)
+                {
+                    throw new Exception("IAsyncParseNodeFactory is required for async operations");
+                }
+
+                return await vendorFactoryAsync.GetRootParseNodeAsync(vendorSpecificContentType, content, cancellationToken).ConfigureAwait(false);
```

Comment:
refactor this body to avoid duplication with the second block

Reply:
Where do you see duplication? I mean it's just duplicated sync code + type cast & check.